### PR TITLE
damlc: incremental package db initialization

### PR DESCRIPTION
**Pull request by robin-da**
### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by robin-da** [comment]
Where's the proper place to add a test for this?

File: compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs [hunk 1]
```diff
@@ -70,120 +72,135 @@ import SdkVersion
...
+
+    -- Read the metadata of an already existing package database and see if wee need to reinitialize.
+    let metadataFileFp = metadataFile projectRoot
+    hasMetadata <- doesFileExist metadataFileFp
+    needsReinitalization <- if hasMetadata
+                            then do
+                              metaData <- readMetadata projectRoot
+                              pure $ hashDependencies metaData /= Just depHash
+                            else pure True
+    when needsReinitalization $ do
```
Comment (cocreature) [inline compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs:96]: This is growing quite long, maybe a good idea to factor out the check as well as the body of the `when`, so this just turns into something like
```
createProjectPackageDb … = do
  needsReinitialization <- getNeedsReinitialization
  when needsReinitialization $ initializeDb
```

Fine to leave that for a separate PR.

File: compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs [hunk 2]
```diff
@@ -70,120 +72,135 @@ import SdkVersion
...
+        foldM
+            (\acc dep -> do
+                 bs <- BSL.readFile dep
+                 pure $ encode $ hash $ bs <> acc)
+            BSL.empty
+            (deps ++ dataDeps)
+
+    -- Read the metadata of an already existing package database and see if wee need to reinitialize.
+    let metadataFileFp = metadataFile projectRoot
+    hasMetadata <- doesFileExist metadataFileFp
```
Comment (cocreature) [inline compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs:90]: `doesFileExist` is almost always racy, I recomend reading it and catching the exception instead. Probably not a big issue here but I’d rather go for the safe version everywhere instead of wondering where the exists check is safe.

File: compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs [hunk 3]
```diff
@@ -70,120 +72,135 @@ import SdkVersion
...
-    writeMetadata projectRoot (PackageDbMetadata (mainUnitIds dependencyInfo) validatedModulePrefixes)
+createProjectPackageDb projectRoot (disableScenarioService -> opts) (PackageSdkVersion thisSdkVer) modulePrefixes deps' dataDeps
+  = do
+    -- compute the hash over all dependencies
+    deps <- expandSdkPackages (optDamlLfVersion opts) (filter (`notElem` basePackages) deps')
+    depHash <-
+        fmap hash $
+        foldM
+            (\acc dep -> do
+                 bs <- BSL.readFile dep
```
Comment (cocreature) [inline compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs:83]: Can we use `https://hackage.haskell.org/package/ghc-8.10.2/docs/Fingerprint.html#v:getFileHash` instead? Afaict, that is specifically written for this type of task and is decently fast and constant memory without depending on lazy IO. We use it in other places already (at least indirectly via ghcide).

File: compiler/damlc/daml-opts/daml-opts-types/DA/Daml/Options/Packaging/Metadata.hs [hunk 4]
```diff
@@ -40,6 +38,9 @@ data PackageDbMetadata = PackageDbMetadata
   -- We do not bother differentiating between exposed and unexposed modules
   -- since we already warn on non-exposed modules anyway and this
   -- is intended for data-dependencies where everything is exposed.
+  , hashDependencies :: Maybe Int
```
Comment (cocreature) [inline compiler/damlc/daml-opts/daml-opts-types/DA/Daml/Options/Packaging/Metadata.hs:41]: Why is this a `Maybe`? Afaict, we always set it to a `Just`.

**Review by cocreature — COMMENTED** [review]
Very nice! This turned out much simpler than I expected. A few minor comments and I’ll take another look once tests are there.

As for tests, I think I’d probably just add a new test suite. In principle, it could go in the packaging tests but those are already really slow and we had to spend some time making them faster recently so I’m hesitant to add it there.

I think there are 4 different tests we should have:

1. Dependency changed and the project depends on that change (e.g., addition of a new function).
2. Dependency added and the project depends on it.
3. Dependency removed as a regression test.
4. Nothing changed, check that modification time hasn’t changed. https://github.com/digital-asset/daml/blob/441f503a95c936832012d9d3c68823b61063a469/compiler/damlc/tests/src/DA/Test/IncrementalBuilds.hs does something similar for `--incremental=yes`.

File: compiler/damlc/daml-opts/daml-opts-types/DA/Daml/Options/Packaging/Metadata.hs (hunk 4 above)
Comment (robin-da) [reply compiler/damlc/daml-opts/daml-opts-types/DA/Daml/Options/Packaging/Metadata.hs:41]: I made it a `Maybe` because otherwise, you'll get an ugly JSON-parsing error if you run the newest SDK version in a project with an already existing package database and metadata file.

File: compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs (hunk 1 above)
Comment (robin-da) [reply compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs:96]: I'll factor it out.

File: compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs (hunk 3 above)
Comment (robin-da) [reply compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs:83]: Sure, I'll use this one instead.

File: compiler/damlc/daml-opts/daml-opts-types/DA/Daml/Options/Packaging/Metadata.hs (hunk 4 above)
Comment (robin-da) [reply compiler/damlc/daml-opts/daml-opts-types/DA/Daml/Options/Packaging/Metadata.hs:41]: But if we catch the exception in `readMetadata`, we can also drop the `Maybe` and reinitialize and rewrite the metadata when it fails reading the existing metadata.

Comment (cocreature) [reply compiler/damlc/daml-opts/daml-opts-types/DA/Daml/Options/Packaging/Metadata.hs:41]: This actually brings up a good point. We need to include the SDK version and the daml-lf version in the hash as well. Catching the exceptions and reinitializing sounds good.

File: compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs (hunk 2 above)
Comment (robin-da) [reply compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs:90]: now catching the exception.

File: compiler/damlc/tests/src/DA/Test/IncrementalPackageDb.hs [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+                , unlines
+                      [ "sdk-version: 0.0.0"
+                      , "name: proj"
+                      , "source: daml"
+                      , "version: 0.0.1"
+                      , "dependencies:"
+                      , "  - daml-prim"
+                      , "  - daml-stdlib"
+                      , "  - dependency/dep.dar"
+                      ])
```
Comment (cocreature) [inline compiler/damlc/tests/src/DA/Test/IncrementalPackageDb.hs:110]: maybe reference something from `B` to make sure everything really worked correctly and we didn’t just write the dep again?

File: compiler/damlc/tests/src/DA/Test/IncrementalPackageDb.hs [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+              ]
+              [ ( "daml/B.daml"
+                , unlines
+                      [ "module B where"
+                      , "template T"
+                      , "  with"
+                      , "    p: Party"
+                      , "  where"
+                      , "    signatory p"
+                      , "f : Int -> Int"
```
Comment (cocreature) [inline compiler/damlc/tests/src/DA/Test/IncrementalPackageDb.hs:80]: Might be nice to make `A` depend on the change in `B`, e.g., also modify it to reference a new definition that we add in `B`.

**Review by cocreature — APPROVED** [review]
Awesome, thank you! Very excited that this long-standing bug is finally fixed.

File: compiler/damlc/tests/src/DA/Test/IncrementalPackageDb.hs [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+                , unlines
+                      [ "module A where"
+                      , "import B"
+                      , "x : Int"
+                      , "x = f 1"])
+              ]
+              []
+              [ ( "daml.yaml"
+                , unlines
+                      [ "sdk-version: 0.0.0"
```
[the remark this answers is not in the document]
Comment (robin-da) [reply compiler/damlc/tests/src/DA/Test/IncrementalPackageDb.hs]: right, I keep forgetting about this. it's fixed now.

File: compiler/damlc/tests/src/DA/Test/IncrementalPackageDb.hs (hunk 5 above)
Comment (robin-da) [reply compiler/damlc/tests/src/DA/Test/IncrementalPackageDb.hs:111]: added a reference.

File: compiler/damlc/tests/src/DA/Test/IncrementalPackageDb.hs (hunk 6 above)
Comment (robin-da) [reply compiler/damlc/tests/src/DA/Test/IncrementalPackageDb.hs:87]: added an additional reference.

File: compiler/damlc/tests/src/DA/Test/IncrementalPackageDb.hs [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+    defaultMain $ tests damlc
+
+tests :: FilePath -> TestTree
+tests damlc =
+    testGroup
+        "Incremental package db initialization"
+        [ test
+              "Reinitialize when dependency changes"
+              [ ( "daml.yaml"
+                , unlines
```
[the remark this answers is not in the document]
Comment (robin-da) [reply compiler/damlc/tests/src/DA/Test/IncrementalPackageDb.hs]: factored it out.
